### ISSUE TITLE
refactor: EKF and KF call closure twice per step; KF wastes H on second call

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#57]
## Problem

Both `ExtendedKalmanFilter.step()` and `KalmanFilter.step()` call `closure()` twice:

1. First call: to get residuals (and Jacobian H) for the update.
2. Second call: after the weight update, just to compute the return-value loss.

For **EKF**, the second call recomputes residuals via autograd, which is fine but redundant — the loss could be computed from the post-update residuals without a full closure re-evaluation.

For **KF**, the second call is worse: the closure must return `(errors, H)`, but `H` is thrown away (`errors, _ = closure()`). The user pays the cost of constructing `H` a second time for no reason.

## Proposal

1. **KF**: Accept an optional `loss_closure` or change the return contract so the second call only needs to return residuals. Alternatively, split the closure into `residual_closure` and `observation_closure`.

2. **EKF**: After updating weights, compute the loss from the post-update residuals directly instead of re-calling the full closure. Since `self.loss()` is just `errors @ errors`, we only need the residuals, not the full autodiff graph.

This would halve the per-step closure evaluations for both optimizers.

**Comment by kovacoj** [comment]
## Proposed fix

**For KalmanFilter:** Change the return logic to avoid re-calling the full closure:

```python
# Before (line 112)
errors, _ = closure()
return self.loss(errors.view(-1))

# After
with torch.no_grad():
    return self.loss(closure().view(-1))
```

But this still requires the user to return `(errors, H)` even though we only need `errors`. A cleaner approach:

**Option A — Loss-only second call:**

Add a `loss_closure` parameter or convention where the second call can return just residuals. This changes the public API, which may not be desirable.

**Option B — Compute loss from residuals directly (preferred):**

Since `self.loss()` is just `errors @ errors`, and the EKF already has `errors` from the post-update Jacobian computation, we can avoid the second closure call entirely:

```python
# In EKF.step(), after weight update:
with torch.no_grad():
    new_errors = closure()
return self.loss(new_errors)
```

Wait — EKF doesn't recompute errors after the update. The simplest fix is:

```python
# After update_weights, just call closure once more under no_grad
with torch.no_grad():
    return self.loss(closure())
```

This still calls closure twice, but the second call is under `no_grad` so it doesn't build a graph — that's the main performance win.

**For the KF H-waste issue specifically**, the cleanest fix is to split the closure contract:

```python
def step(self, closure=None, observation_closure=None):
    errors, H = closure()  # full closure for the update
    ...
    # For the return value, only need residuals
    return self.loss(observation_closure() if observation_closure else closure()[0])
```

But this changes the public API. A non-breaking alternative: just wrap the second call in `torch.no_grad()` so at least the H computation doesn't build a graph:

```python
with torch.no_grad():
    errors, _ = closure()
    return self.loss(errors.view(-1))
```

**Recommendation:** Start with the `torch.no_grad()` wrapper for both EKF and KF second calls — zero API change, immediate perf win. The `observation_closure` split can be a separate enhancement.
